datafusion/optimizer/src/rewrite_cycle.rs

```diff
@@ -0,0 +1,262 @@
+    /// calls and then record the number at the end of the first cycle.
+    fn record_cycle_length(&mut self) {
+        self.cycle_length = Some(self.rewrite_count);
+    }
+
+    /// Returns true when the loop has reached the maximum cycle length or when we've received
+    /// consecutive unchanged tree nodes equal to the total number of rewriters.
+    fn is_done(&self) -> bool {
+        // default value indicates we have not completed a cycle
+        let Some(cycle_length) = self.cycle_length else {
+            return false;
+        };
+        self.consecutive_unchanged_count >= cycle_length
+    }
+
+    /// Finishes the iteration by consuming the state and returning a [TreeNode] and
+    /// [RewriteCycleInfo]
+    fn finish(self) -> Result<(Node, RewriteCycleInfo)> {
+        Ok((
+            self.node,
+            RewriteCycleInfo {
+                cycle_length: self.cycle_length.unwrap_or(self.rewrite_count),
+                total_iterations: self.rewrite_count,
+            },
+        ))
+    }
+
+    /// Calls [TreeNode::rewrite] and determines if the rewrite cycle should break or continue
+    /// based on the current [RewriteCycleState].
+    pub fn rewrite<R: TreeNodeRewriter<Node = Node> + ?Sized>(
+        mut self,
+        rewriter: &mut R,
+    ) -> RewriteCycleControlFlow<Self> {
+        match self.node.rewrite(rewriter) {
+            Err(e) => ControlFlow::Break(Err(e)),
+            Ok(Transformed {
+                data: node,
+                transformed,
+                ..
+            }) => {
+                self.node = node;
+                self.rewrite_count += 1;
+                if transformed {
+                    self.consecutive_unchanged_count = 0;
+                } else {
+                    self.consecutive_unchanged_count += 1;
+                }
+                if self.is_done() {
+                    ControlFlow::Break(Ok(self))
+                } else {
+                    ControlFlow::Continue(self)
+                }
+            }
+        }
+    }
+}
+
+/// Information about a rewrite cycle, such as total number of iterations and number of fully
+/// completed cycles. This is useful for testing purposes to ensure that optimzation passes are
+/// working as expected.
+#[derive(Debug, Clone, Copy)]
+pub struct RewriteCycleInfo {
+    total_iterations: usize,
+    cycle_length: usize,
+}
+
+impl RewriteCycleInfo {
+    /// The total number of **fully completed** cycles.
+    pub fn completed_cycles(&self) -> usize {
+        self.total_iterations / self.cycle_length
+    }
+
+    /// The total number of [TreeNode::rewrite] calls.
+    pub fn total_iterations(&self) -> usize {
+        self.total_iterations
+    }
+
+    /// The number of [TreeNode::rewrite] calls within a single cycle.
+    pub fn cycle_length(&self) -> usize {
+        self.cycle_length
+    }
+}
+
+pub type RewriteCycleControlFlow<T> = ControlFlow<Result<T>, T>;
```

Comment:
Since you have this API so nicely broken out, I think it would be really neat to add tests for it that demonstrate it works properly in isolation as well as make its api clearer in code